g11n-ws/modules/md-restful-api-i18n/src/main/java/com/vmware/vip/i18n/api/base/StreamProductAction.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2019-2023 VMware, Inc.
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package com.vmware.vip.i18n.api.base;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.vmware.vip.common.constants.ConstantsChar;
+import com.vmware.vip.common.constants.ConstantsKeys;
+import com.vmware.vip.common.constants.ConstantsMsg;
+import com.vmware.vip.common.i18n.status.APIResponseStatus;
+import com.vmware.vip.core.messages.exception.L3APIException;
+import com.vmware.vip.core.messages.service.multcomponent.IMultComponentService;
+import com.vmware.vip.core.messages.service.product.IProductService;
+import com.vmware.vip.messages.data.dao.model.ResultMessageChannel;
+
+public class StreamProductAction extends TranslationProductAction{
+	 private static byte[] byteComm = (ConstantsChar.COMMA+"\r\n").getBytes();
+	
+	@Autowired
+	IProductService productService;
+	
+	@Autowired
+	IMultComponentService multComponentService;
+	
+	
+	public void writeMultTranslationResponse(String productName, String version, String components, String locales,
+			String pseudo, HttpServletResponse resp) throws Exception {
+
+		String newVersion = super.getAvailableVersion(productName, version);
+		List<String> componentList = getcomponentList(productName, newVersion, components);
+		List<String> localeList = getlocaleList(productName, newVersion, locales, pseudo);
+
+		List<ResultMessageChannel>  readChannels = multComponentService.getTranslationChannels(productName, newVersion, componentList, localeList);
+		if(readChannels.isEmpty()) {
+            throw new L3APIException(ConstantsMsg.TRANS_IS_NOT_FOUND);
+        }
+
+		StreamProductResp sr = new StreamProductResp(productName, newVersion, localeList, componentList, pseudo, false);
+		resp.setContentType(ConstantsKeys.CONTENT_TYPE_JSON);
+		WritableByteChannel wbc = Channels.newChannel(resp.getOutputStream());
+
+		boolean isPartContent = writeResponseHeader(version.equals(newVersion), componentList.size() * localeList.size(), readChannels.size(), wbc, sr);
+		writeTranslationsToChanel(readChannels, wbc);
+		 if(isPartContent){
+			 wbc.write(ByteBuffer.wrap(addNullMessage(readChannels, componentList, localeList)));
+		 }
+		 wbc.write(sr.getEndBytes());
+	}
+
+	private boolean writeResponseHeader(boolean notVersionFallback, int expSize, int readChannelSize, WritableByteChannel wbc, StreamProductResp sr) throws IOException {
```

Comment:
What does it mean by `Header`?

Reply:
write the response header bytes to channel